src/toolkit/matl_buy_policy.h

```diff
@@ -128,15 +141,18 @@ class MatlBuyPolicy : public Trader {
   /// is idempotent.
   void Stop();
 
+  double sample_fraction_ = 1.0;
+
   /// the total amount requested
   inline double TotalQty() const {
-    return std::min(throughput_,
-                    fill_to_ * buf_->capacity() - buf_->quantity());
+    return sample_fraction_ * std::min(throughput_,
+           (fill_to_ * buf_->capacity()) - buf_->quantity());
```

Comment:
I think you just want to move the logic/code from `SetRequestSize()` into `TotalQty()` since that code is needed if and only if `TotalQty()` is called.